internal/distro/image_config.go

```diff
@@ -62,98 +67,31 @@ type ImageConfig struct {
-			finalConfig.DefaultTarget = parentConfig.DefaultTarget
-		}
-		if finalConfig.Sysconfig == nil {
-			finalConfig.Sysconfig = parentConfig.Sysconfig
-		}
-		if finalConfig.GPGKeyFiles == nil {
-			finalConfig.GPGKeyFiles = parentConfig.GPGKeyFiles
-		}
-		if finalConfig.RHSMConfig == nil {
-			finalConfig.RHSMConfig = parentConfig.RHSMConfig
-		}
-		if finalConfig.SystemdLogind == nil {
-			finalConfig.SystemdLogind = parentConfig.SystemdLogind
-		}
-		if finalConfig.CloudInit == nil {
-			finalConfig.CloudInit = parentConfig.CloudInit
-		}
-		if finalConfig.Modprobe == nil {
-			finalConfig.Modprobe = parentConfig.Modprobe
-		}
-		if finalConfig.DracutConf == nil {
-			finalConfig.DracutConf = parentConfig.DracutConf
-		}
-		if finalConfig.SystemdUnit == nil {
-			finalConfig.SystemdUnit = parentConfig.SystemdUnit
-		}
-		if finalConfig.Authselect == nil {
-			finalConfig.Authselect = parentConfig.Authselect
-		}
-		if finalConfig.SELinuxConfig == nil {
-			finalConfig.SELinuxConfig = parentConfig.SELinuxConfig
-		}
-		if finalConfig.Tuned == nil {
-			finalConfig.Tuned = parentConfig.Tuned
-		}
-		if finalConfig.Tmpfilesd == nil {
-			finalConfig.Tmpfilesd = parentConfig.Tmpfilesd
-		}
-		if finalConfig.PamLimitsConf == nil {
-			finalConfig.PamLimitsConf = parentConfig.PamLimitsConf
-		}
-		if finalConfig.Sysctld == nil {
-			finalConfig.Sysctld = parentConfig.Sysctld
-		}
-		if finalConfig.DNFConfig == nil {
-			finalConfig.DNFConfig = parentConfig.DNFConfig
-		}
-		if finalConfig.SshdConfig == nil {
-			finalConfig.SshdConfig = parentConfig.SshdConfig
-		}
-		if finalConfig.Authconfig == nil {
-			finalConfig.Authconfig = parentConfig.Authconfig
-		}
-		if finalConfig.PwQuality == nil {
-			finalConfig.PwQuality = parentConfig.PwQuality
-		}
-		if finalConfig.DNFAutomaticConfig == nil {
-			finalConfig.DNFAutomaticConfig = parentConfig.DNFAutomaticConfig
-		}
-		if finalConfig.YumConfig == nil {
-			finalConfig.YumConfig = parentConfig.YumConfig
-		}
-		if finalConfig.YUMRepos == nil {
-			finalConfig.YUMRepos = parentConfig.YUMRepos
-		}
-		if finalConfig.Firewall == nil {
-			finalConfig.Firewall = parentConfig.Firewall
-		}
-		if finalConfig.UdevRules == nil {
-			finalConfig.UdevRules = parentConfig.UdevRules
-		}
-		if finalConfig.GCPGuestAgentConfig == nil {
-			finalConfig.GCPGuestAgentConfig = parentConfig.GCPGuestAgentConfig
+		// iterate over all struct fields and copy unset values from the parent
+		for i := 0; i < reflect.TypeOf(*c).NumField(); i++ {
+			fieldName := reflect.TypeOf(*c).Field(i).Name
+			field := reflect.ValueOf(&finalConfig).Elem().FieldByName(fieldName)
+			setValue := false
+			switch field.Kind() {
+			case reflect.Ptr:
+				if field.IsNil() {
+					setValue = true
+				}
+			case reflect.Slice:
+				if field.IsNil() {
+					setValue = true
+				}
+			case reflect.Map:
+				if field.IsNil() {
+					setValue = true
+				}
+			default:
+				panic(fmt.Sprintf("unsupported field type: %s", field.Kind()))
```

Comment:
To avoid concrete types being added in the future, maybe we should add a comment or expand the error message to add something like "use pointers instead of concrete types".